src/Lecture2.hs

```diff
@@ -268,7 +401,23 @@ data EvalError
 It returns either a successful evaluation result or an error.
 -}
 eval :: Variables -> Expr -> Either EvalError Int
-eval = error "TODO"
+eval _ (Lit n) = Right n
+eval l (Add a b) =
+  let
+   expRes1 = eval l a
+   expRes2 = eval l b
+  in sumExpRes expRes1 expRes2
+eval l (Var s) =
+  let
+    varVal = lookup s l
+  in
+    maybe (Left (VariableNotFound s)) Right varVal
+
+sumExpRes :: Either EvalError Int -> Either EvalError Int -> Either EvalError Int
+sumExpRes expRes1 expRes2
+    | isLeft expRes1 = expRes1
+    | isLeft expRes2 = expRes2
+    | otherwise = Right (fromRight 0 expRes1 + fromRight 0 expRes2)
```

Comment:
Using `isLeft` + `fromRight` is called a "boolean blindness" antipattern. You can use pattern matching here instead:

```suggestion
sumExpRes (Left e) _ = Left e
sumExpRes _ (Left e) = Left e
sumExpRes (Right x) (Right y) = Right (x + y)
```

Once you learn about Monads, you'll be able to reduce this function to a single line ✂️